src/main/java/net/snowflake/ingest/streaming/internal/DataValidationUtil.java

```diff
@@ -303,6 +300,93 @@ static String validateAndParseObject(String columnName, Object input) {
+  private static OffsetDateTime inputToOffsetDateTime(
+      String columnName, String typeName, Object input, ZoneId defaultTimezone) {
+    if (input instanceof OffsetDateTime) {
+      return (OffsetDateTime) input;
+    }
+
+    if (input instanceof ZonedDateTime) {
+      return ((ZonedDateTime) input).toOffsetDateTime();
+    }
+
+    if (input instanceof LocalDateTime) {
+      return ((LocalDateTime) input).atZone(defaultTimezone).toOffsetDateTime();
+    }
+
+    if (input instanceof LocalDate) {
+      return ((LocalDate) input).atStartOfDay().atZone(defaultTimezone).toOffsetDateTime();
+    }
+
+    if (input instanceof Instant) {
+      // Just like integer-stored timestamps, instants are always interpreted in UTC
+      return ((Instant) input).atZone(ZoneOffset.UTC).toOffsetDateTime();
+    }
+
+    if (input instanceof String) {
+      String stringInput = (String) input;
+      {
+        // First, try to parse ZonedDateTime
+        ZonedDateTime zoned = catchParsingError(() -> ZonedDateTime.parse(stringInput));
+        if (zoned != null) return zoned.toOffsetDateTime();
+      }
+
+      {
+        // Next, try to parse OffsetDateTime
+        OffsetDateTime offset = catchParsingError(() -> OffsetDateTime.parse(stringInput));
+        if (offset != null) return offset;
+      }
+
+      {
+        // Alternatively, try to parse LocalDateTime
+        LocalDateTime localDateTime = catchParsingError(() -> LocalDateTime.parse(stringInput));
+        if (localDateTime != null) return localDateTime.atZone(defaultTimezone).toOffsetDateTime();
+      }
+
+      {
+        // Alternatively, try to parse LocalDate
+        LocalDate localDate = catchParsingError(() -> LocalDate.parse(stringInput));
+        if (localDate != null)
+          return localDate.atStartOfDay().atZone(defaultTimezone).toOffsetDateTime();
+      }
+
+      {
+        // Alternatively, try to parse integer-stored timestamp
+        // Just like in Snowflake, integer-stored timestamps are always in UTC
+        Instant instant = catchParsingError(() -> parseInstantGuessScale(stringInput));
+        if (instant != null) return instant.atOffset(ZoneOffset.UTC);
+      }
+
+      // Couldn't parse anything, throw an exception
+      throw valueFormatNotAllowedException(
+          columnName,
+          input.toString(),
+          typeName,
+          "Not a valid value, see"
+              + " https://docs.snowflake.com/en/LIMITEDACCESS/snowpipe-streaming.html"
+              + " for the list of supported formats");
+    }
+
+    // Type is not supported, throw an exception
+    throw typeNotAllowedException(
+        columnName,
+        input.getClass(),
+        typeName,
+        new String[] {"String", "LocalDate", "LocalDateTime", "ZonedDateTime", "OffsetDateTime"});
+  }
+
+  private static <T> T catchParsingError(Supplier<T> op) {
+    try {
+      return op.get();
+    } catch (DateTimeParseException | NumberFormatException e) {
+      return null;
```

Comment:
Is this right? Should we throw an exception here?

Reply:
This is right, it is just helper to translate an exception to null. The reason is that we are trying to parse input in 5 different formats, any of them can fail and I wanted to avoid 5 nested try/catch blocks.

Reply:
Ack, makes sense.